### PR TITLE
Adding Token Merge Interceptor

### PR DESCRIPTION
**Pull request by CrowleyRajapakse**
## Purpose

- Implementing Token merge interceptor in which two parts of the divided token coming in the header and in cookie, will be merged and sent as a single token as the "Authorization" token.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/interceptors/auth/TokenMergeInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    private static final Log logger = LogFactory.getLog(TokenMergeInterceptor.class);
+
+    public TokenMergeInterceptor() {
+        //We will use PRE_INVOKE phase as we need to process message before hit actual service
+        super(Phase.PRE_INVOKE);
+    }
+
+    public void handleMessage(Message message) throws Fault {
+        //If Authorization headers are present anonymous URI check will be skipped
+
```
Comment (shilmyhasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/interceptors/auth/TokenMergeInterceptor.java]: additional space between the comment

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/interceptors/auth/TokenMergeInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import org.apache.cxf.interceptor.Fault;
+import org.apache.cxf.message.Message;
+import org.apache.cxf.phase.AbstractPhaseInterceptor;
+import org.apache.cxf.phase.Phase;
+import org.wso2.carbon.apimgt.rest.api.util.RestApiConstants;
+
+import java.util.ArrayList;
+import java.util.TreeMap;
+
+public class TokenMergeInterceptor extends AbstractPhaseInterceptor {
```
Comment (isharac) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/interceptors/auth/TokenMergeInterceptor.java]: Please add java doc comments

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/interceptors/auth/TokenMergeInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        TreeMap headers = (TreeMap) message.get(Message.PROTOCOL_HEADERS);
+        if (authHeaders != null || tokenCookie != null) {
+            String headerTokenString = authHeaders.get(0).toString();
+            String cookieTokenString = tokenCookie.get(0).toString();
+            int startingIndexOfToken =
+                    cookieTokenString.lastIndexOf(RestApiConstants.AUTH_COOKIE_NAME) + RestApiConstants.AUTH_COOKIE_NAME
+                            .length() + 1;
+            String tokenStringPartTwo = cookieTokenString.substring(startingIndexOfToken);
+            String accessToken = headerTokenString + tokenStringPartTwo;
+            ArrayList AuthorizationHeader = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/interceptors/auth/TokenMergeInterceptor.java]: corrected in the latest code

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/interceptors/auth/TokenMergeInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        //If Authorization headers are present anonymous URI check will be skipped
+
+        ArrayList authHeaders = (ArrayList) ((TreeMap) (message.get(Message.PROTOCOL_HEADERS)))
+                .get(RestApiConstants.AUTH_HEADER_NAME);
+
+        ArrayList tokenCookie = (ArrayList) ((TreeMap) (message.get(Message.PROTOCOL_HEADERS)))
+                .get(RestApiConstants.COOKIE_NAME);
+
+        TreeMap headers = (TreeMap) message.get(Message.PROTOCOL_HEADERS);
+        if (authHeaders != null || tokenCookie != null) {
```
[the remark this answers is not in the document]
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/interceptors/auth/TokenMergeInterceptor.java]: outdated code

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/interceptors/auth/TokenMergeInterceptor.java (hunk 2 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/interceptors/auth/TokenMergeInterceptor.java:30]: resolved by #8332

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/interceptors/auth/TokenMergeInterceptor.java (hunk 1 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/interceptors/auth/TokenMergeInterceptor.java:41]: resolved by #8332

**Pull request merged**